include/macros.h

```diff
@@ -276,6 +276,19 @@ extern struct GraphicsContext* __gfxCtx;
 
 // HackerOoT
 
+// assertf, print a message before abort
+#if IS_DEBUG
+# define assertf(cond, fmt, ...)                    \
+    do {                                            \
+        if (!(cond)) {                              \
+            PRINTF(fmt, __VA_ARGS__);               \
+            __assert(#cond, __FILE__, __LINE__);    \
+        }                                           \
+    } while (0)
+#else
+# define assertf(cond, fmt, ...) ((void)0)
+#endif
+
```

Comment:
This is essentially vanilla asserts with a console message attached. It's just a convenience macro for providing some extra information only when the assert is triggered.

Reply:
Okay, I think it would be best if new debugging systems were not introduced in unrelated PRs, but if there was really nothing in vanilla that covers this use case, go ahead.

Reply:
It's here because I needed it to debug this PR, it feels counterproductive to remove it after the fact especially if there are still obscure double-frees lurking in vanilla code since they're a pain to debug without the extra info this macro provides. If we're really concerned about clean history, I can make a new PR and stall this one until that one makes it in.